hail/src/main/resources/include/hail/Utils.h

```diff
@@ -133,6 +133,34 @@ class ArrayLoadImpl : public BaseArrayImpl<elem_required, elem_size, elem_align>
   }
 };
 
+template<typename ElemT, size_t elem_size, size_t elem_align>
```

Comment:
The hand-layout in other classes is to match the layout between JVM and C++, but NDArray is C++ native, so you don't need to bother with that.  You can just make this a C++ class, like:

```
class NDArray {
public:
  long flag;
  std::vector<long> shape;
  std::vector<long> offset;
  std::vector<long> strides;
  char *data;

  // constructors, methods, etc.
};
```